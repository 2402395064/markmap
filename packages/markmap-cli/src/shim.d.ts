declare interface Window {
  mm: import('markmap-view').Markmap;
  markmap: typeof import('markmap-toolbar');
}
