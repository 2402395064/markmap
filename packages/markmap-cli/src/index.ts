export * from 'markmap-lib';
export * from './dev-server';
