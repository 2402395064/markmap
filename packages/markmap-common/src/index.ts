export * from './hook';
export * from './html';
export * from './util';

export * from './types';
