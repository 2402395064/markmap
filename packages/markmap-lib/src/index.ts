import { promises as fs } from 'fs';
import { transform, getUsedAssets } from './transform';
import { fillTemplate } from './template';
import { IMarkmapCreateOptions } from './types';

export * from './types';

export * from './template';
export * from './transform';

export async function createMarkmap(options: IMarkmapCreateOptions = {}): Promise<string> {
  const { root, features } = transform(options.content || '');
  const assets = getUsedAssets(features);
  const html = fillTemplate(root, assets);
  const output = options.output || 'markmap.html';
  await fs.writeFile(output, html, 'utf8');
  return output;
}
