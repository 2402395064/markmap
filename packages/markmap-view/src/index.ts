import * as d3 from 'd3';
import { flextree } from 'd3-flextree';
import {
  INode,
  Hook,
  getId,
  walkTree,
  arrayFrom,
  addClass,
  childSelector,
  noop,
  loadJS,
  loadCSS,
} from 'markmap-common';
import {
  IMarkmapOptions,
  IMarkmapState,
  IMarkmapFlexTreeItem,
  IMarkmapLinkItem,
} from './types';
import css from './style.css';
import containerCSS from './container.css';

export const globalCSS = css;

export { loadJS, loadCSS };
export * from './types';

interface IPadding {
  left: number;
  right: number;
  top: number;
  bottom: number;
}

function linkWidth(nodeData: IMarkmapFlexTreeItem): number {
  const data: INode = nodeData.data;
  return Math.max(6 - 2 * data.depth, 1.5);
}

function adjustSpacing(tree: IMarkmapFlexTreeItem, spacing: number): void {
  walkTree(
    tree,
    (d, next) => {
      d.ySizeInner = d.ySize - spacing;
      d.y += spacing;
      next();
    },
    'children'
  );
}

function minBy(numbers: number[], by: (v: number) => number): number {
  const index = d3.minIndex(numbers, by);
  return numbers[index];
}

function stopPropagation(e: Event) {
  e.stopPropagation();
}

type ID3SVGElement = d3.Selection<
  SVGElement,
  IMarkmapFlexTreeItem,
  HTMLElement,
  IMarkmapFlexTreeItem
>;

function createViewHooks() {
  return {
    transformHtml: new Hook<[mm: Markmap, nodes: HTMLElement[]]>(),
  };
}

/**
 * A global hook to refresh all markmaps when called.
 */
export const refreshHook = new Hook<[]>();

export const defaultColorFn = d3.scaleOrdinal(d3.schemeCategory10);

const isMacintosh =
  typeof navigator !== 'undefined' && navigator.userAgent.includes('Macintosh');

export class Markmap {
  static defaultOptions: IMarkmapOptions = {
    duration: 500,
    embedGlobalCSS: true,
    nodeMinHeight: 16,
    spacingVertical: 5,
    spacingHorizontal: 80,
    autoFit: false,
    fitRatio: 0.95,
    color: (node: INode): string => defaultColorFn(`${node.state.id}`),
    paddingX: 8,
    scrollForPan: isMacintosh,
  };

  options: IMarkmapOptions;

  state: IMarkmapState;

  svg: ID3SVGElement;

  styleNode: d3.Selection<
    HTMLStyleElement,
    IMarkmapFlexTreeItem,
    HTMLElement,
    IMarkmapFlexTreeItem
  >;

  g: d3.Selection<
    SVGGElement,
    IMarkmapFlexTreeItem,
    HTMLElement,
    IMarkmapFlexTreeItem
  >;

  zoom: d3.ZoomBehavior<Element, unknown>;

  viewHooks: ReturnType<typeof createViewHooks>;

  revokers: (() => void)[] = [];

  constructor(
    svg: string | SVGElement | ID3SVGElement,
    opts?: Partial<IMarkmapOptions>
  ) {
    ['handleZoom', 'handleClick', 'handlePan'].forEach((key) => {
      this[key] = this[key].bind(this);
    });
    this.viewHooks = createViewHooks();
    this.svg = (svg as ID3SVGElement).datum
      ? (svg as ID3SVGElement)
      : d3.select(svg as string);
    this.styleNode = this.svg.append('style');
    this.zoom = d3
      .zoom()
      .filter((event) => {
        if (this.options.scrollForPan) {
          // Pan with wheels, zoom with ctrl+wheels
          if (event.type === 'wheel') return event.ctrlKey && !event.button;
        }
        return (!event.ctrlKey || event.type === 'wheel') && !event.button;
      })
      .on('zoom', this.handleZoom);
    this.setOptions(opts);
    this.state = {
      id: this.options.id || this.svg.attr('id') || getId(),
    };
    this.g = this.svg.append('g');
    this.updateStyle();
    this.svg.call(this.zoom).on('wheel', this.handlePan);
    this.revokers.push(
      refreshHook.tap(() => {
        this.setData();
      })
    );
  }

  getStyleContent(): string {
    const { style } = this.options;
    const { id } = this.state;
    const styleText = typeof style === 'function' ? style(id) : '';
    return [this.options.embedGlobalCSS && css, styleText]
      .filter(Boolean)
      .join('\n');
  }

  updateStyle(): void {
    this.svg.attr(
      'class',
      addClass(this.svg.attr('class'), 'markmap', this.state.id)
    );
    const style = this.getStyleContent();
    this.styleNode.text(style);
  }

  handleZoom(e) {
    const { transform } = e;
    this.g.attr('transform', transform);
  }

  handlePan(e: WheelEvent) {
    e.preventDefault();
    const transform = d3.zoomTransform(this.svg.node());
    const newTransform = transform.translate(-e.deltaX, -e.deltaY);
    this.svg.call(this.zoom.transform, newTransform);
  }

  handleClick(_: MouseEvent, d: IMarkmapFlexTreeItem): void {
    const { data } = d;
    data.payload = {
      ...data.payload,
      fold: !data.payload?.fold,
    };
    this.renderData(d.data);
  }

  initializeData(node: INode): void {
    let nodeId = 0;
    const { color, nodeMinHeight } = this.options;
    const { id } = this.state;
    const container = document.createElement('div');
    container.className = addClass(
      container.className,
      'markmap',
      'markmap-container',
      'markmap-foreign',
      `${id}-g`
    );
    const style = document.createElement('style');
    style.textContent = [this.getStyleContent(), containerCSS].join('\n');
    document.body.append(style, container);
    walkTree(node, (item, next) => {
      item.children = item.children?.map((child) => ({ ...child }));
      nodeId += 1;
      const el = document.createElement('div');
      el.innerHTML = item.content;
      container.append(el);
      item.state = {
        ...item.state,
        id: nodeId,
        el,
      };
      color(item); // preload colors
      next();
    });
    const nodes = arrayFrom(container.childNodes) as HTMLElement[];
    this.viewHooks.transformHtml.call(this, nodes);
    walkTree(node, (item, next, parent) => {
      const rect = item.state.el.getBoundingClientRect();
      item.content = item.state.el.innerHTML;
      item.state.size = [
        Math.ceil(rect.width),
        Math.max(Math.ceil(rect.height), nodeMinHeight),
      ];
      item.state.path = [parent?.state?.path, item.state.id]
        .filter(Boolean)
        .join('.');
      item.state.key =
        [parent?.state?.id, item.state.id].filter(Boolean).join('.') +
        // FIXME: find a way to check content hash
        item.content;
      next();
    });
    container.remove();
    style.remove();
  }

  setOptions(opts: Partial<IMarkmapOptions>): void {
    this.options = {
      ...Markmap.defaultOptions,
      ...opts,
    };
  }

  setData(data?: INode, opts?: Partial<IMarkmapOptions>): void {
    if (data) this.state.data = data;
    this.initializeData(this.state.data);
    if (opts) this.setOptions(opts);
    this.renderData();
  }

  renderData(originData?: INode): void {
    if (!this.state.data) return;
    const { spacingHorizontal, paddingX, spacingVertical, autoFit, color } =
      this.options;
    const layout = flextree()
      .children((d: INode) => !d.payload?.fold && d.children)
      .nodeSize((d: IMarkmapFlexTreeItem) => {
        const [width, height] = d.data.state.size;
        return [height, width + (width ? paddingX * 2 : 0) + spacingHorizontal];
      })
      .spacing((a: IMarkmapFlexTreeItem, b: IMarkmapFlexTreeItem) => {
        return a.parent === b.parent ? spacingVertical : spacingVertical * 2;
      });
    const tree = layout.hierarchy(this.state.data);
    layout(tree);
    adjustSpacing(tree, spacingHorizontal);
    const descendants: IMarkmapFlexTreeItem[] = tree.descendants().reverse();
    const links: IMarkmapLinkItem[] = tree.links();
    const linkShape = d3.linkHorizontal();
    const minX = d3.min(descendants, (d) => d.x - d.xSize / 2);
    const maxX = d3.max(descendants, (d) => d.x + d.xSize / 2);
    const minY = d3.min(descendants, (d) => d.y);
    const maxY = d3.max(descendants, (d) => d.y + d.ySizeInner);
    Object.assign(this.state, {
      minX,
      maxX,
      minY,
      maxY,
    });

    if (autoFit) this.fit();

    const origin =
      (originData && descendants.find((item) => item.data === originData)) ||
      (tree as IMarkmapFlexTreeItem);
    const x0 = origin.data.state.x0 ?? origin.x;
    const y0 = origin.data.state.y0 ?? origin.y;

    // Update the nodes
    const node = this.g
      .selectAll<SVGGElement, IMarkmapFlexTreeItem>(
        childSelector<SVGGElement>('g')
      )
      .data(descendants, (d) => d.data.state.key);
    const nodeEnter = node
      .enter()
      .append('g')
      .attr('class', 'markmap-node')
      .attr('data-depth', (d) => d.data.depth)
      .attr('data-path', (d) => d.data.state.path)
      .attr(
        'transform',
        (d) =>
          `translate(${y0 + origin.ySizeInner - d.ySizeInner},${
            x0 + origin.xSize / 2 - d.xSize
          })`
      );

    const nodeExit = this.transition(node.exit<IMarkmapFlexTreeItem>());
    nodeExit
      .select('line')
      .attr('x1', (d) => d.ySizeInner)
      .attr('x2', (d) => d.ySizeInner);
    nodeExit.select('foreignObject').style('opacity', 0);
    nodeExit
      .attr(
        'transform',
        (d) =>
          `translate(${origin.y + origin.ySizeInner - d.ySizeInner},${
            origin.x + origin.xSize / 2 - d.xSize
          })`
      )
      .remove();

    const nodeMerge = node.merge(nodeEnter);
    this.transition(nodeMerge).attr(
      'transform',
      (d) => `translate(${d.y},${d.x - d.xSize / 2})`
    );

    // Update lines under the content
    const line = nodeMerge
      .selectAll<SVGLineElement, IMarkmapFlexTreeItem>(
        childSelector<SVGLineElement>('line')
      )
      .data(
        (d) => [d],
        (d) => d.data.state.key
      )
      .join(
        (enter) => {
          return enter
            .append('line')
            .attr('x1', (d) => d.ySizeInner)
            .attr('y1', (d) => d.xSize)
            .attr('x2', (d) => d.ySizeInner)
            .attr('y2', (d) => d.xSize);
        },
        (update) => update,
        (exit) => exit.remove()
      );
    this.transition(line)
      .attr('x1', -1)
      .attr('x2', (d) => d.ySizeInner + 2)
      .attr('stroke', (d) => color(d.data))
      .attr('stroke-width', linkWidth);

    // Circle to link to children of the node
    const circle = nodeMerge
      .selectAll<SVGCircleElement, IMarkmapFlexTreeItem>(
        childSelector<SVGCircleElement>('circle')
      )
      .data(
        (d) => (d.data.children ? [d] : []),
        (d) => d.data.state.key
      )
      .join(
        (enter) => {
          return enter
            .append('circle')
            .attr('stroke-width', '1.5')
            .attr('cx', (d) => d.ySizeInner)
            .attr('cy', (d) => d.xSize)
            .attr('r', 0)
            .on('click', this.handleClick);
        },
        (update) => update,
        (exit) => exit.remove()
      );
    this.transition(circle)
      .attr('r', 6)
      .attr('stroke', (d) => color(d.data))
      .attr('fill', (d) =>
        d.data.payload?.fold && d.data.children ? color(d.data) : '#fff'
      );

    const foreignObject = nodeMerge
      .selectAll<SVGForeignObjectElement, IMarkmapFlexTreeItem>(
        childSelector<SVGForeignObjectElement>('foreignObject')
      )
      .data(
        (d) => [d],
        (d) => d.data.state.key
      )
      .join(
        (enter) => {
          const fo = enter
            .append('foreignObject')
            .attr('class', 'markmap-foreign')
            .attr('x', paddingX)
            .attr('y', 0)
            .style('opacity', 0)
            .attr('height', (d) => d.xSize)
            .on('mousedown', stopPropagation)
            .on('dblclick', stopPropagation);
          fo.append<HTMLDivElement>('xhtml:div')
            .select(function select(d) {
              const clone = d.data.state.el.cloneNode(true) as HTMLElement;
              this.replaceWith(clone);
              return clone;
            })
            .attr('xmlns', 'http://www.w3.org/1999/xhtml');
          return fo;
        },
        (update) => update,
        (exit) => exit.remove()
      )
      .attr('width', (d) => Math.max(0, d.ySizeInner - paddingX * 2));
    this.transition(foreignObject).style('opacity', 1);

    // Update the links
    const path = this.g
      .selectAll<SVGPathElement, IMarkmapLinkItem>(
        childSelector<SVGPathElement>('path')
      )
      .data(links, (d) => d.target.data.state.key)
      .join(
        (enter) => {
          const source: [number, number] = [
            y0 + origin.ySizeInner,
            x0 + origin.xSize / 2,
          ];
          return enter
            .insert('path', 'g')
            .attr('class', 'markmap-link')
            .attr('data-depth', (d) => d.target.data.depth)
            .attr('data-path', (d) => d.target.data.state.path)
            .attr('d', linkShape({ source, target: source }));
        },
        (update) => update,
        (exit) => {
          const source: [number, number] = [
            origin.y + origin.ySizeInner,
            origin.x + origin.xSize / 2,
          ];
          return this.transition(exit)
            .attr('d', linkShape({ source, target: source }))
            .remove();
        }
      );
    this.transition(path)
      .attr('stroke', (d) => color(d.target.data))
      .attr('stroke-width', (d) => linkWidth(d.target))
      .attr('d', (d) => {
        const source: [number, number] = [
          d.source.y + d.source.ySizeInner,
          d.source.x + d.source.xSize / 2,
        ];
        const target: [number, number] = [
          d.target.y,
          d.target.x + d.target.xSize / 2,
        ];
        return linkShape({ source, target });
      });

    descendants.forEach((d) => {
      d.data.state.x0 = d.x;
      d.data.state.y0 = d.y;
    });
  }

  transition<T extends d3.BaseType, U, P extends d3.BaseType, Q>(
    sel: d3.Selection<T, U, P, Q>
  ): d3.Transition<T, U, P, Q> {
    const { duration } = this.options;
    return sel.transition().duration(duration);
  }

  /**
   * Fit the content to the viewport.
   */
  async fit(): Promise<void> {
    const svgNode = this.svg.node();
    const { width: offsetWidth, height: offsetHeight } =
      svgNode.getBoundingClientRect();
    const { fitRatio } = this.options;
    const { minX, maxX, minY, maxY } = this.state;
    const naturalWidth = maxY - minY;
    const naturalHeight = maxX - minX;
    const scale = Math.min(
      (offsetWidth / naturalWidth) * fitRatio,
      (offsetHeight / naturalHeight) * fitRatio,
      2
    );
    const initialZoom = d3.zoomIdentity
      .translate(
        (offsetWidth - naturalWidth * scale) / 2 - minY * scale,
        (offsetHeight - naturalHeight * scale) / 2 - minX * scale
      )
      .scale(scale);
    return this.transition(this.svg)
      .call(this.zoom.transform, initialZoom)
      .end()
      .catch(noop);
  }

  /**
   * Pan the content to make the provided node visible in the viewport.
   */
  async ensureView(
    node: INode,
    padding: Partial<IPadding> | undefined
  ): Promise<void> {
    let g: SVGGElement | undefined;
    let itemData: IMarkmapFlexTreeItem | undefined;
    this.g
      .selectAll<SVGGElement, IMarkmapFlexTreeItem>(
        childSelector<SVGGElement>('g')
      )
      .each(function walk(d) {
        if (d.data === node) {
          g = this;
          itemData = d;
        }
      });
    if (!g || !itemData) return;
    const svgNode = this.svg.node();
    const relRect = svgNode.getBoundingClientRect();
    const transform = d3.zoomTransform(svgNode);
    const [left, right] = [
      itemData.y,
      itemData.y + itemData.ySizeInner + 2,
    ].map((x) => x * transform.k + transform.x);
    const [top, bottom] = [
      itemData.x - itemData.xSize / 2,
      itemData.x + itemData.xSize / 2,
    ].map((y) => y * transform.k + transform.y);
    // Skip if the node includes or is included in the container.
    const pd: IPadding = {
      left: 0,
      right: 0,
      top: 0,
      bottom: 0,
      ...padding,
    };
    const dxs = [pd.left - left, relRect.width - pd.right - right];
    const dys = [pd.top - top, relRect.height - pd.bottom - bottom];
    const dx = dxs[0] * dxs[1] > 0 ? minBy(dxs, Math.abs) / transform.k : 0;
    const dy = dys[0] * dys[1] > 0 ? minBy(dys, Math.abs) / transform.k : 0;
    if (dx || dy) {
      const newTransform = transform.translate(dx, dy);
      return this.transition(this.svg)
        .call(this.zoom.transform, newTransform)
        .end()
        .catch(noop);
    }
  }

  /**
   * Scale content with it pinned at the center of the viewport.
   */
  async rescale(scale: number): Promise<void> {
    const svgNode = this.svg.node();
    const { width: offsetWidth, height: offsetHeight } =
      svgNode.getBoundingClientRect();
    const halfWidth = offsetWidth / 2;
    const halfHeight = offsetHeight / 2;
    const transform = d3.zoomTransform(svgNode);
    const newTransform = transform
      .translate(
        ((halfWidth - transform.x) * (1 - scale)) / transform.k,
        ((halfHeight - transform.y) * (1 - scale)) / transform.k
      )
      .scale(scale);
    return this.transition(this.svg)
      .call(this.zoom.transform, newTransform)
      .end()
      .catch(noop);
  }

  destroy() {
    this.svg.on('.zoom', null);
    this.svg.html(null);
    this.revokers.forEach((fn) => {
      fn();
    });
  }

  static create(
    svg: string | SVGElement | ID3SVGElement,
    opts?: Partial<IMarkmapOptions>,
    data?: INode
  ): Markmap {
    const mm = new Markmap(svg, opts);
    if (data) {
      mm.setData(data);
      mm.fit(); // always fit for the first render
    }
    return mm;
  }
}
