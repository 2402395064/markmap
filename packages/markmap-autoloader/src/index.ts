import { buildJSItem, buildCSSItem, loadCSS, loadJS } from 'markmap-common';
import type { AutoLoaderOptions } from './types';

const enabled: Record<string, boolean> = {};

const autoLoaderOptions: AutoLoaderOptions = {
  baseJs: [
    `d3@${process.env.D3_VERSION}`,
    `markmap-lib@${process.env.LIB_VERSION}`,
    `markmap-view@${process.env.VIEW_VERSION}`,
    `markmap-toolbar@${process.env.TOOLBAR_VERSION}`,
  ],
  baseCss: [`markmap-toolbar@${process.env.TOOLBAR_VERSION}/dist/style.css`],
  manual: false,
  toolbar: false,
  ...window.markmap?.autoLoader,
};

export const ready = Promise.all([
  loadJS(
    autoLoaderOptions.baseJs.map((item) =>
      typeof item === 'string' ? buildJSItem(item) : item
    )
  ),
  loadCSS(
    autoLoaderOptions.baseCss.map((item) =>
      typeof item === 'string' ? buildCSSItem(item) : item
    )
  ),
]).then(() => {
  const { markmap } = window;
  const style = document.createElement('style');
  style.textContent = markmap.globalCSS;
  // Insert global CSS to body so it has higher priority than prism.css, etc.
  document.body.prepend(style);
  autoLoaderOptions.onReady?.();
});

function transform(
  transformer: import('markmap-lib').Transformer,
  content: string
) {
  const result = transformer.transform(content);
  const keys = Object.keys(result.features).filter((key) => !enabled[key]);
  keys.forEach((key) => {
    enabled[key] = true;
  });
  const { styles, scripts } = transformer.getAssets(keys);
  const { markmap } = window;
  if (styles) markmap.loadCSS(styles);
  if (scripts) markmap.loadJS(scripts);
  return result;
}

export function render(el: HTMLElement) {
  const { Transformer, Markmap, deriveOptions, Toolbar } = window.markmap;
  const lines = el.textContent.split('\n');
  let indent = Infinity;
  lines.forEach((line) => {
    const spaces = line.match(/^\s*/)[0].length;
    if (spaces < line.length) indent = Math.min(indent, spaces);
  });
  const content = lines
    .map((line) => line.slice(indent))
    .join('\n')
    .trim();
  const transformer = new Transformer(autoLoaderOptions.transformPlugins);
  el.innerHTML = '<svg></svg>';
  const svg = el.firstChild as SVGElement;
  const mm = Markmap.create(svg, { embedGlobalCSS: false });
  if (autoLoaderOptions.toolbar) {
    const { el: toolbar } = Toolbar.create(mm);
    Object.assign(toolbar.style, {
      position: 'absolute',
      right: '20px',
      bottom: '20px',
    });
    el.append(toolbar);
  }
  const doRender = () => {
    const { root, frontmatter } = transform(transformer, content);
    const markmapOptions = frontmatter?.markmap;
    const frontmatterOptions = deriveOptions(markmapOptions);
    mm.setData(root, frontmatterOptions);
    mm.fit();
  };
  transformer.hooks.retransform.tap(doRender);
  doRender();
}

export async function renderAllUnder(container: ParentNode) {
  await ready;
  container.querySelectorAll('.markmap').forEach(render);
}

export function renderAll() {
  return renderAllUnder(document);
}

if (!autoLoaderOptions.manual) {
  if (document.readyState === 'loading')
    document.addEventListener('DOMContentLoaded', () => {
      renderAll();
    });
  else renderAll();
}
