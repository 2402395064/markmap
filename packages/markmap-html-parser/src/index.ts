import { Cheerio, Element, load } from 'cheerio';
import { IPureNode, walkTree } from 'markmap-common';

export enum Levels {
  None,
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
  Block,
  List,
  ListItem,
}

export interface IHtmlNode {
  id: number;
  tag: string;
  html: string;
  level: Levels;
  parent: number;
  childrenLevel: Levels;
  children?: IHtmlNode[];
  comments?: string[];
}

export interface IHtmlParserOptions {
  /** Matches nodes to be displayed on markmap. */
  selector: string;
  /** Matches wrapper elements that should be ignored. */
  selectorWrapper: string;
  /** Matches elements that can be nested, such as `li`. */
  selectorNesting: string;
  /** Matches elements whose tag name should be preserved, such as `<pre>`, `<img>`. */
  selectorPreserveTag: string;
}

const MARKMAP_COMMENT_PREFIX = 'markmap: ';
const SELECTOR_HEADING = /^h[1-6]$/;
const SELECTOR_LIST = /^[uo]l$/;
const SELECTOR_LIST_ITEM = /^li$/;

export const defaultOptions: IHtmlParserOptions = {
  selector: 'h1,h2,h3,h4,h5,h6,ul,ol,li,table,pre,p>img:first-child:last-child',
  selectorWrapper: 'div,p',
  selectorNesting: 'ul,ol,li',
  selectorPreserveTag: 'table,pre,img',
};

function getLevel(tagName: string) {
  if (SELECTOR_HEADING.test(tagName)) return +tagName[1] as Levels;
  if (SELECTOR_LIST.test(tagName)) return Levels.List;
  if (SELECTOR_LIST_ITEM.test(tagName)) return Levels.ListItem;
  return Levels.Block;
}

export function parseHtml(html: string, opts?: Partial<IHtmlParserOptions>) {
  const options = {
    ...defaultOptions,
    ...opts,
  };
  const $ = load(html);
  const $root = $('body');
  let id = 0;
  const rootNode: IHtmlNode = {
    id,
    tag: '',
    html: '',
    level: Levels.None,
    parent: 0,
    childrenLevel: Levels.None,
    children: [],
  };
  const elMap: Record<number, Cheerio<Element>> = {
    [id]: $<Element, string>('<div>'),
  };

  const headingStack: IHtmlNode[] = [];
  let skippingHeading = Levels.None;
  checkNode($root);

  walkTree(rootNode, (node, next) => {
    next();
    const $el = elMap[node.id];
    // Extract magic comments
    $el.contents().each((_, child) => {
      if (child.type === 'comment') {
        const data = child.data.trim();
        if (data.startsWith(MARKMAP_COMMENT_PREFIX)) {
          node.comments ||= [];
          node.comments.push(data.slice(MARKMAP_COMMENT_PREFIX.length).trim());
          $(child).remove();
        }
      }
    });
    node.html = (
      ($el.is(options.selectorPreserveTag) ? $.html($el) : $el.html()) || ''
    ).trimEnd();
  });
  return rootNode;

  function addChild(
    $child: Cheerio<Element>,
    parent: IHtmlNode,
    nesting: boolean,
  ) {
    const child = $child[0];
    const node: IHtmlNode = {
      id: ++id,
      tag: child.tagName,
      html: '',
      level: getLevel(child.tagName),
      childrenLevel: Levels.None,
      children: nesting ? [] : undefined,
      parent: parent.id,
    };
    if (parent.children) {
      if (
        parent.childrenLevel === Levels.None ||
        parent.childrenLevel > node.level
      ) {
        parent.children = [];
        parent.childrenLevel = node.level;
      }
      if (parent.childrenLevel === node.level) {
        parent.children.push(node);
      }
    }
    elMap[node.id] = $child;
    return node;
  }

  function getCurrentHeading(level: Levels) {
    let heading: IHtmlNode | undefined;
    while ((heading = headingStack.at(-1)) && heading.level >= level) {
      headingStack.pop();
    }
    return heading || rootNode;
  }

  function checkNode($el: Cheerio<Element>, node?: IHtmlNode) {
    $el.children().each((_, child) => {
      const $child = $(child);
      if ($child.is(options.selectorWrapper)) {
        checkNode($child);
        return;
      }
      if (!$child.is(options.selector)) {
        const level = getLevel(child.tagName);
        if (level <= Levels.H6) {
          skippingHeading = level;
        }
        return;
      }
      const level = getLevel(child.tagName);
      if (skippingHeading > Levels.None && level > skippingHeading) {
        return;
      }
      skippingHeading = Levels.None;
      if ($child.is(options.selectorNesting)) {
        const childNode = addChild(
          $child,
          node || getCurrentHeading(level),
          true,
        );
        if (childNode) {
          checkNode($child, childNode);
          $child.remove();
        }
        return;
      }
      const isHeading = level <= Levels.H6;
      const childNode = addChild(
        $child,
        node || getCurrentHeading(level),
        isHeading,
      );
      if (isHeading) headingStack.push(childNode);
    });
  }
}

export function convertNode(htmlRoot: IHtmlNode) {
  return walkTree<IHtmlNode, IPureNode>(htmlRoot, (htmlNode, next) => {
    const node: IPureNode = {
      content: htmlNode.html,
      children: next() || [],
    };
    if (htmlNode.comments) {
      if (htmlNode.comments.includes('foldAll')) {
        node.payload = { fold: 2 };
      } else if (htmlNode.comments.includes('fold')) {
        node.payload = { fold: 1 };
      }
    }
    return node;
  });
}

export function buildTree(html: string, opts?: Partial<IHtmlParserOptions>) {
  const htmlRoot = parseHtml(html, opts);
  return convertNode(htmlRoot);
}
