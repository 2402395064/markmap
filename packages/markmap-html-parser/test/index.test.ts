import { parseHtml, convertNode } from '../src/index';

test('parseHtml', () => {
  const root = parseHtml(`
<body>
<div class="container">
<h1 data-id="h1">heading 1</h1>
<p>text here will be ignored.</p>
<ul>
<li>this list</li>
<li>is ignored</li>
</ul>
<h2 data-id="h2">heading 2 <!-- markmap: foldAll --></h2>
<p>text also ignored</p>
<ul>
<li><p>item 1</p></li>
<li><p>item 2</p>
<ul>
<li>item 3</li>
<li>item 4</li>
</li>
</ul>
</ul>
<ol>
<li>item 5</li>
<li>item 6</li>
</ol>
<h2>A table</h2>
<table>
<tr>
<td>
<ul><li>this list is ignored</li></ul>
</td>
</tr>
</table>
<p><img src="image1.png"></p>
<p><img src="image2.png"></p>
</div>
</body>
`);
  expect(root).toMatchSnapshot();
  expect(convertNode(root)).toMatchSnapshot();
});
