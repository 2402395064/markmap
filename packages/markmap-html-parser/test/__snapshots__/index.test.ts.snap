// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseHtml 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "childrenLevel": 0,
                  "html": "<p>item 1</p>",
                  "id": 7,
                  "level": 9,
                  "parent": 6,
                  "tag": "li",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "childrenLevel": 0,
                          "html": "item 3",
                          "id": 10,
                          "level": 9,
                          "parent": 9,
                          "tag": "li",
                        },
                        {
                          "children": [],
                          "childrenLevel": 0,
                          "html": "item 4",
                          "id": 11,
                          "level": 9,
                          "parent": 9,
                          "tag": "li",
                        },
                      ],
                      "childrenLevel": 9,
                      "html": "",
                      "id": 9,
                      "level": 8,
                      "parent": 8,
                      "tag": "ul",
                    },
                  ],
                  "childrenLevel": 8,
                  "html": "<p>item 2</p>",
                  "id": 8,
                  "level": 9,
                  "parent": 6,
                  "tag": "li",
                },
              ],
              "childrenLevel": 9,
              "html": "",
              "id": 6,
              "level": 8,
              "parent": 5,
              "tag": "ul",
            },
            {
              "children": [
                {
                  "children": [],
                  "childrenLevel": 0,
                  "html": "item 5",
                  "id": 13,
                  "level": 9,
                  "parent": 12,
                  "tag": "li",
                },
                {
                  "children": [],
                  "childrenLevel": 0,
                  "html": "item 6",
                  "id": 14,
                  "level": 9,
                  "parent": 12,
                  "tag": "li",
                },
              ],
              "childrenLevel": 9,
              "html": "",
              "id": 12,
              "level": 8,
              "parent": 5,
              "tag": "ol",
            },
          ],
          "childrenLevel": 8,
          "comments": [
            "foldAll",
          ],
          "html": "heading 2",
          "id": 5,
          "level": 2,
          "parent": 1,
          "tag": "h2",
        },
        {
          "children": [
            {
              "children": undefined,
              "childrenLevel": 0,
              "html": "<table>
<tbody><tr>
<td>
<ul><li>this list is ignored</li></ul>
</td>
</tr>
</tbody></table>",
              "id": 16,
              "level": 7,
              "parent": 15,
              "tag": "table",
            },
            {
              "children": undefined,
              "childrenLevel": 0,
              "html": "<img src="image1.png">",
              "id": 17,
              "level": 7,
              "parent": 15,
              "tag": "img",
            },
            {
              "children": undefined,
              "childrenLevel": 0,
              "html": "<img src="image2.png">",
              "id": 18,
              "level": 7,
              "parent": 15,
              "tag": "img",
            },
          ],
          "childrenLevel": 7,
          "html": "A table",
          "id": 15,
          "level": 2,
          "parent": 1,
          "tag": "h2",
        },
      ],
      "childrenLevel": 2,
      "html": "heading 1",
      "id": 1,
      "level": 1,
      "parent": 0,
      "tag": "h1",
    },
  ],
  "childrenLevel": 1,
  "html": "",
  "id": 0,
  "level": 0,
  "parent": 0,
  "tag": "",
}
`;

exports[`parseHtml 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "content": "<p>item 1</p>",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "content": "item 3",
                        },
                        {
                          "children": [],
                          "content": "item 4",
                        },
                      ],
                      "content": "",
                    },
                  ],
                  "content": "<p>item 2</p>",
                },
              ],
              "content": "",
            },
            {
              "children": [
                {
                  "children": [],
                  "content": "item 5",
                },
                {
                  "children": [],
                  "content": "item 6",
                },
              ],
              "content": "",
            },
          ],
          "content": "heading 2",
          "payload": {
            "fold": 2,
          },
        },
        {
          "children": [
            {
              "children": [],
              "content": "<table>
<tbody><tr>
<td>
<ul><li>this list is ignored</li></ul>
</td>
</tr>
</tbody></table>",
            },
            {
              "children": [],
              "content": "<img src="image1.png">",
            },
            {
              "children": [],
              "content": "<img src="image2.png">",
            },
          ],
          "content": "A table",
        },
      ],
      "content": "heading 1",
    },
  ],
  "content": "",
}
`;
